torch_geometric/nn/aggr/scaler.py

```diff
@@ -0,0 +1,74 @@
+from typing import Any, Dict, List, Optional, Union
+
+import torch
+from torch import Tensor
+
+from torch_geometric.nn.aggr import Aggregation, MultiAggregation
+from torch_geometric.utils import degree
+
+
+class DegreeScalerAggregation(Aggregation):
+    """
+    Class that combines together one or more aggregators and then transforms
+    the result with one or more scalers. The scalers are normalised by the
+    in-degree of the training set and so must be provided at construction.
+
+    Args:
+        aggr (string or list or Aggregation, optional): The list of
+            aggregations given as :class:`~torch_geometric.nn.aggr.Aggregation`
+            (or any string that automatically resolves to it).
+        scalers (list of str): Set of scaling function identifiers, namely
+                :obj:`"identity"`, :obj:`"amplification"`,
+                :obj:`"attenuation"`, :obj:`"linear"` and
+                :obj:`"inverse_linear"`.
+        deg (Tensor): Histogram of in-degrees of nodes in the training set,
+            used by scalers to normalize.
+        aggr_kwargs (List[Dict[str, Any]], optional): Arguments passed to the
+            respective aggregation functions in case it gets automatically
+            resolved. (default: :obj:`None`)
+    """
+    def __init__(self, aggrs: List[Union[Aggregation, str]],
+                 scalers: List[str], deg: Tensor,
+                 aggrs_kwargs: Optional[List[Dict[str, Any]]] = None):
+
+        super().__init__()
+
+        self.agg = MultiAggregation(aggrs, aggrs_kwargs)
+        self.scalers = scalers
+
+        deg = deg.to(torch.float)
+        num_nodes = int(deg.sum())
+        bin_degrees = torch.arange(deg.numel())
+        self.avg_deg: Dict[str, float] = {
+            'lin': float((bin_degrees * deg).sum()) / num_nodes,
+            'log': float(((bin_degrees + 1).log() * deg).sum()) / num_nodes,
+            'exp': float((bin_degrees.exp() * deg).sum()) / num_nodes,
+        }
+
+    def forward(self, x: Tensor, index: Optional[Tensor] = None,
+                ptr: Optional[Tensor] = None, dim_size: Optional[int] = None,
+                dim: int = -2) -> Tensor:
+
+        # TODO: Does not currently support ptr because `degree` doesn't.
+        if index is None:
+            raise NotImplementedError("ptr not currently supported")
+
+        out = self.agg(x, index, ptr, dim_size, dim)
+        deg = degree(index, dtype=out.dtype)
+        deg = deg.clamp_(1).view(*([-1] + [1] * (len(out.shape) - 1)))
```

Comment:
This  was modified - before it was simply 
```
deg.clamp_(1).view(-1, 1, 1)
```
I don't know if  there is a much cleaner way to achieve the same  thing here.

Reply:
Thanks for fixing. I think this is still not completely right though. How about we take `out.size()` and write -1 to the dimension given to `dim`, and use that for reshaping.

Reply:
Hmm makes sense to use the dimension give to `dim`, but  just to clarify on using `out.size()`, if for example 
```
out.size() = [4, 4, 96]
deg.size() = [4]
dim = -2
```
I assume we want to do
```
deg.view(1, -1, 1)
```
so that the scaler can broadcast

Does that make sense?

Reply:
Yes, that‘s what I meant. Sorry for the confusion.